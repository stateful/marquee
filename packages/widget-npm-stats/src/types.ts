import type { ContextProperties } from '@vscode-marquee/utils'

export type StatResponse = {
  [packageName: string]: Record<string, number>
}

export interface State {
  stats: StatResponse
  isLoading: boolean
  error?: Error | null
}
export interface Configuration {
  packageNames: string[]
  from?: number
  until?: number
}

export interface Context extends ContextProperties<Configuration & State> { }

type JSONValue =
  | string
  | number
  | boolean
  | JSONObject
  | JSONArray

export interface JSONObject {
  [x: string]: JSONValue
}

export interface JSONArray extends Array<JSONValue> { }
