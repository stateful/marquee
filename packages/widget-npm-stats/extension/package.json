{
  "name": "@vscode-marquee/widget-npm-stats-extension",
  "description": "Extension logic for NPM Stats Widget",
  "main": "../build/extension.js",
  "module": "../build/extension.js",
  "types": "../build/extension.d.ts"
}
