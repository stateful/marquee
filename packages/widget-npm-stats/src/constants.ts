import type { Configuration, State } from './types'

const ONE_YEAR = 1000 * 60 * 60 * 24 * 365
export const DEFAULT_CONFIGURATION: Configuration = {
  packageNames: [],
  from: Date.now() - ONE_YEAR,
  until: Date.now(),
}
export const DEFAULT_STATE: State = {
  stats: {},
  isLoading: true
}

export const STATS_URL = 'https://npm-stat.com/api/download-counts'
export const LEGEND_HEIGHT = 80
export const MIN_HEIGHT = 300
export const TEXT_COLOR = 'var(--vscode-editorWidget-foreground)'
