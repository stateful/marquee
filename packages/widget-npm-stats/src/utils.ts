const LOCALE_DATE_OPTIONS = {
  year: 'numeric',
  month: '2-digit',
  day: '2-digit',
} as const

export function formatDate (ts: number) {
  return (new Date(ts))
    .toLocaleDateString('en-GB', LOCALE_DATE_OPTIONS)
    .split('/')
    .reverse()
    .join('-')
}
