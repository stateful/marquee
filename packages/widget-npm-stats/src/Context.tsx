import React, { createContext, useContext } from 'react'
import { connect, getEventListener, MarqueeWindow } from '@vscode-marquee/utils'

import type { Context, Configuration, State } from './types'

declare const window: MarqueeWindow
const NPMStatsContext = createContext<Context>({} as Context)
export const WIDGET_ID = '@vscode-marquee/npm-stats-widget'

export const NPMStatsProvider = ({ children }: { children: React.ReactElement }) => {
  const widgetState = getEventListener<Configuration & State>(WIDGET_ID)
  const providerValues = connect<Configuration & State>({
    ...window.marqueeStateConfiguration[WIDGET_ID].configuration,
    ...window.marqueeStateConfiguration[WIDGET_ID].state
  }, widgetState)

  return (
    <NPMStatsContext.Provider value={{ ...providerValues }}>
      {children}
    </NPMStatsContext.Provider>
  )
}

export const useNPMStatsContext = () => {
  return useContext(NPMStatsContext)
}

export default NPMStatsContext
