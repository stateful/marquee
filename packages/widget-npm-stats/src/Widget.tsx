import React, { useContext, useRef, useEffect, useState } from 'react'
import { Grid, Typography, CircularProgress, Button } from '@mui/material'
import styled from '@emotion/styled'
import 'react-vis/dist/style.css'

import wrapper, { Dragger, HeaderWrapper } from '@vscode-marquee/widget'
import { NetworkError, MarqueeWindow } from '@vscode-marquee/utils'
import type { MarqueeWidgetProps } from '@vscode-marquee/widget'

import PopMenu from './components/Pop'
import NPMGraph from './components/Graph'
import NPMStatsContext, { NPMStatsProvider } from './Context'
import { TEXT_COLOR, MIN_HEIGHT, LEGEND_HEIGHT } from './constants'

declare const window: MarqueeWindow

const CENTER_STYLES = {
  overflow: 'auto',
  height: '100%',
  width: '100%',
  padding: '24px',
}

const Root = styled(Grid)(() => ({
  ['.rv-discrete-color-legend-item__title']: {
    color: TEXT_COLOR
  },
  ['.rv-discrete-color-legend.horizontal']: {
    textAlign: 'center'
  },
  b: {
    fontWeight: 'bold'
  }
}))

const WidgetBody = () => {
  const ref = useRef<HTMLDivElement>(null)
  const { error, isLoading, stats } = useContext(NPMStatsContext)
  const [ graphHeight, setGraphHeight ] = useState<number>(MIN_HEIGHT)
  const hasStats = Object.keys(stats).length !== 0

  useEffect(() => {
    setGraphHeight(ref.current
      ? ref.current.clientHeight - LEGEND_HEIGHT
      : MIN_HEIGHT
    )
  }, [ref.current?.clientHeight])

  return (
    <Grid item xs ref={ref}>
      <Grid
        container
        wrap="nowrap"
        direction="column"
        style={{ height: '100%' }}
      >
        {isLoading && (
          <Grid
            container
            style={{ height: '100%' }}
            alignItems="center"
            justifyContent="center"
            direction="column"
            data-testid="loading"
          >
            <Grid item>
              <CircularProgress color="secondary" />
            </Grid>
          </Grid>
        )}
        {error && !isLoading && (
          <Grid item xs style={CENTER_STYLES} data-testid="error">
            <NetworkError message={error.message} />
          </Grid>
        )}
        {!error && !isLoading && !hasStats && (
          <Grid item xs style={{ ...CENTER_STYLES, textAlign: 'center' }} data-testid="no-stats">
            No packages defined in your Marquee configuration!<br />
            <Button style={{ marginTop: 10 }} onClick={() => window.vscode.postMessage({
              west: { execCommands: [{
                command: 'workbench.action.openSettings',
                args: ['@ext:stateful.marquee Npm-stats Package Names']
              }]},
            })}>
              Add NPM Packages Now!
            </Button>
          </Grid>
        )}
        {!error && !isLoading && hasStats && (
          <Grid
            item
            xs
            style={{
              overflow: 'auto',
              paddingTop: '16px',
              paddingRight: '16px',
              paddingLeft: '16px',
              paddingBottom: '8px',
            }}
          >
            <Grid
              container
              style={{
                height: '100%',
                width: '100%',
              }}
              direction="column"
              wrap="nowrap"
            >
              <Root item xs={1} style={{ maxWidth: '100%' }}>
                <NPMGraph height={graphHeight} stats={stats} />
              </Root>
            </Grid>
          </Grid>
        )}
      </Grid>
    </Grid>
  )
}

let Welcome = ({ ToggleFullScreen }: MarqueeWidgetProps) => {
  return (
    <>
      <HeaderWrapper>
        <Grid item>
          <Typography variant="subtitle1">NPM Statistics</Typography>
        </Grid>
        <Grid item>
          <Grid container direction="row" spacing={1}>
            <Grid item>
              <PopMenu />
            </Grid>
            <Grid item>
              <ToggleFullScreen />
            </Grid>
            <Grid item>
              <Dragger />
            </Grid>
          </Grid>
        </Grid>
      </HeaderWrapper>
      <WidgetBody />
    </>
  )
}

const Widget = (props: any) => (
  <NPMStatsProvider>
    <Welcome {...props} />
  </NPMStatsProvider>
)
export default wrapper(Widget, 'welcome')
