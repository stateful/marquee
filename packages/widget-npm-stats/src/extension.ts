import vscode from 'vscode'
import { format } from 'util'
import Axios from 'axios'
import ExtensionManager, { Logger, ChildLogger } from '@vscode-marquee/utils/extension'
import type { Client } from 'tangle'

import { formatDate } from './utils'
import { DEFAULT_CONFIGURATION, DEFAULT_STATE, STATS_URL } from './constants'
import type { Configuration, State, JSONObject, StatResponse } from './types'

const STATE_KEY = 'widgets.npm-stats'
export class NPMStatsExtensionManager extends ExtensionManager<State, Configuration> {
  #logger: ChildLogger
  private _isFetching = false

  constructor (context: vscode.ExtensionContext) {
    super(context, STATE_KEY, DEFAULT_CONFIGURATION, DEFAULT_STATE)
    this.#logger = Logger.getChildLogger(STATE_KEY)
    this._checkWorkspaceForNPMPackage()
  }

  private async _checkWorkspaceForNPMPackage () {
    /**
     * don't check for NPM packages if
     */
    if (
      /**
       * no workspace was opened
       */
      !vscode.workspace.workspaceFolders ||
      /**
       * there are custom packages defined the VS Code configuration
       */
      this._configuration.packageNames.length !== 0
    ) {
      return
    }

    try {
      const workspacePath = vscode.workspace.workspaceFolders[0].uri
      const pkgJson = vscode.Uri.joinPath(workspacePath, 'package.json')
      const pkgJsonContent: JSONObject = JSON.parse((await vscode.workspace.fs.readFile(pkgJson)).toString())
      if (pkgJsonContent.name) {
        this.#logger.info(`Detected NPM workspace, adding ${pkgJsonContent.name as string} to the configuration`)
        await this.updateConfiguration(
          'packageNames',
          [pkgJsonContent.name as string],
          vscode.ConfigurationTarget.Workspace
        )
        this.broadcast({ packageNames: [pkgJsonContent.name as string] })
      }
    } catch (err: any) {
      return
    }
  }

  private async _loadStatistics () {
    if (this._isFetching) {
      return
    }

    this._isFetching = true
    await this.updateState('isLoading', this._isFetching)
    if (this.configuration.packageNames.length === 0) {
      await this.updateState('stats', {})
      await this.updateState('error', null)
      this._isFetching = false
      await this.updateState('isLoading', this._isFetching)
      this._tangle?.broadcast({
        stats: {},
        isLoading: false,
        error: null
      } as State & Configuration)
      return
    }

    try {
      /**
       * check if from is set after until and throw an error before making the request
       */
      if (
        this.configuration.from &&
        this.configuration.until &&
        (this.configuration.from > this.configuration.until)
      ) {
        throw new Error('The "from" time configuration can not be set after the date of "until"')
      }

      const packageParam = this.configuration.packageNames
        .map((p) => `package=${p}`)
        .join('&')
      const url = format(
        '%s?%s&from=%s&until=%s',
        STATS_URL,
        packageParam,
        formatDate(this.configuration.from || DEFAULT_CONFIGURATION.from),
        formatDate(this.configuration.until || DEFAULT_CONFIGURATION.until)
      )

      this.#logger.info(`Fetch NPM Stats from ${url}`)
      const result = await Axios.get(url).then(
        (res) => res.data as StatResponse,
        (err) => err as Error
      )

      if (result instanceof Error) {
        throw new Error(`Fetching NPM stats failed: ${(result as Error).message}`)
      }

      await this.updateState('stats', result)
      await this.updateState('error', null)
      this._isFetching = false
      await this.updateState('isLoading', this._isFetching)

      this._tangle?.broadcast({
        stats: result,
        isLoading: false,
        error: null
      } as State & Configuration)
    } catch (err: any) {
      this._isFetching = false
      await this.updateState('isLoading', this._isFetching)
      await this.updateState('error', { message: err.message } as Error)
      this._tangle?.broadcast({
        isLoading: false,
        error: { message: err.message } as Error
      } as State & Configuration)
      this.#logger.error(err.message)
    }
  }

  public setBroadcaster (tangle: Client<State & Configuration>) {
    super.setBroadcaster(tangle)
    this.on('configurationUpdate', () => this._loadStatistics())
    this._loadStatistics()
    return this
  }
}

export function activate (context: vscode.ExtensionContext) {
  const stateManager = new NPMStatsExtensionManager(context)
  return {
    marquee: {
      disposable: stateManager,
      defaultState: stateManager.state,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    },
  }
}
