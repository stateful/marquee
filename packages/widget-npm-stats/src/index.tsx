import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBarChart } from '@fortawesome/free-solid-svg-icons/faBarChart'

import Widget from './Widget'
import { NPMStatsProvider } from './Context'

export default {
  name: 'npm-stats',
  icon: <FontAwesomeIcon icon={faBarChart} />,
  label: 'NPM Statistics',
  tags: ['package', 'npm', 'stats', 'download', 'statistics'],
  description: 'Widget to show download statistics for NPM package',
  component: Widget,
}
export { Widget, NPMStatsProvider }
