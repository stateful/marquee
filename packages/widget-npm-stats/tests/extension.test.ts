import { request } from 'undici'
import { activate } from '../src/extension'

const channel = {
  appendLine: jest.fn()
}

describe('Extension Manager', () => {
  let manager: any

  beforeEach(() => {
    manager = activate({} as any, channel as any)
    ;(manager.marquee.disposable.updateState as jest.Mock).mockClear()
  })

  it('should initiate extension manager properly', async () => {
    await manager.marquee.disposable['_loadStatistics']()
    expect(manager.marquee.disposable.updateState).toBeCalledTimes(4)
    expect(manager.marquee.disposable.updateState).toBeCalledWith('isLoading', true)
    expect(manager.marquee.disposable.updateState).toBeCalledWith('stats', {})
    expect(manager.marquee.disposable.updateState).toBeCalledWith('error', null)
    expect(manager.marquee.disposable.updateState).toBeCalledWith('isLoading', false)
  })

  it('should successfully fetch data', async () => {
    manager.marquee.disposable.configuration.packageNames = ['foo', 'bar']
    await manager.marquee.disposable['_loadStatistics']()
    expect((manager.marquee.disposable.updateState as jest.Mock).mock.calls)
      .toMatchSnapshot()
  })

  it('should fail properly', async () => {
    (request as jest.Mock).mockRejectedValue(new Error('ups'))
    await manager.marquee.disposable['_loadStatistics']()
    expect((manager.marquee.disposable.updateState as jest.Mock).mock.calls)
      .toMatchSnapshot()
  })
})
