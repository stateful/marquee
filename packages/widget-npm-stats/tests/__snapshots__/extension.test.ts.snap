// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Extension Manager should fail properly 1`] = `
Array [
  Array [
    "isLoading",
    true,
  ],
  Array [
    "isLoading",
    false,
  ],
  Array [
    "error",
    Object {
      "message": "ups",
    },
  ],
]
`;

exports[`Extension Manager should successfully fetch data 1`] = `
Array [
  Array [
    "isLoading",
    true,
  ],
  Array [
    "stats",
    Object {
      "bar": Object {
        "2022-02-01": 33334,
        "2022-02-02": 31277,
      },
      "foo": Object {
        "2022-06-26": 45142,
        "2022-06-27": 169454,
      },
    },
  ],
  Array [
    "isLoading",
    false,
  ],
  Array [
    "error",
    null,
  ],
]
`;
